disk_store.go

```diff
@@ -98,47 +100,83 @@ func NewDiskStore(fileName string) (*DiskStore, error) {
 	return ds, nil
 }
 
-func (d *DiskStore) Get(key string) string {
+func (d *DiskStore) Get(key string) (string, error) {
 	// Get retrieves the value from the disk and returns. If the key does not
 	// exist then it returns an empty string
 	//
 	// How get works?
 	//	1. Check if there is any KeyEntry record for the key in keyDir
-	//	2. Return an empty string if key doesn't exist
+	//	2. Return an empty string if key doesn't exist or if the key has been deleted
 	//	3. If it exists, then read KeyEntry.totalSize bytes starting from the
 	//     KeyEntry.position from the disk
 	//	4. Decode the bytes into valid KV pair and return the value
 	//
 	kEntry, ok := d.keyDir[key]
 	if !ok {
-		return ""
+		return "", ErrKeyNotFound
 	}
+
 	// move the current pointer to the right offset
-	// TODO: handle errors
-	d.file.Seek(int64(kEntry.position), defaultWhence)
+	_, err := d.file.Seek(int64(kEntry.position), defaultWhence)
+	if err != nil {
+		return "", ErrSeekFailed
+	}
+
 	data := make([]byte, kEntry.totalSize)
-	// TODO: handle errors
-	_, err := io.ReadFull(d.file, data)
+	_, err = io.ReadFull(d.file, data)
 	if err != nil {
-		panic("read error")
+		return "", ErrReadFailed
 	}
-	_, _, value := decodeKV(data)
-	return value
+
+	result := &Record{}
+	result.DecodeKV(data)
+	return result.Value, nil
 }
 
-func (d *DiskStore) Set(key string, value string) {
+func (d *DiskStore) Set(key string, value string) error {
 	// Set stores the key and value on the disk
 	//
 	// The steps to save a KV to disk is simple:
 	// 1. Encode the KV into bytes
 	// 2. Write the bytes to disk by appending to the file
 	// 3. Update KeyDir with the KeyEntry of this key
 	timestamp := uint32(time.Now().Unix())
-	size, data := encodeKV(timestamp, key, value)
-	d.write(data)
-	d.keyDir[key] = NewKeyEntry(timestamp, uint32(d.writePosition), uint32(size))
+	h := Header{TimeStamp: timestamp, KeySize: uint32(len(key)), ValueSize: uint32(len(value))}
+	r := Record{Header: h, Key: key, Value: value}
+
+	//encode the record
+	buf := bytes.NewBuffer(make([]byte, headerSize))
+	err := r.EncodeKV(buf)
+	if err != nil {
+		log.Fatalf("error in encoding the value %v", err)
+		return ErrEncodingFailed
+	}
+	d.write(buf.Bytes())
+	size := headerSize + h.KeySize + h.ValueSize
```

Comment:
how about making size as property of `Record`, like `Record.Size()`? It has all the necessary information